### PR TITLE
Support for final metering reports [DPP-819]

### PR DESCRIPTION
**Pull request by simonmaxen-da**

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiMeteringReportService.scala [hunk 1]
```diff
@@ -85,25 +86,30 @@ private[apiserver] object ApiMeteringReportService {
   }
 
   def toTimestamp(ts: ProtoTimestamp): Either[String, Timestamp] = {
-    Timestamp.fromInstant(Instant.ofEpochSecond(ts.seconds, ts.nanos.toLong))
+    val utcTs =
+      OffsetDateTime.ofInstant(Instant.ofEpochSecond(ts.seconds, ts.nanos.toLong), ZoneOffset.UTC)
+    for {
+      _ <-
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiMeteringReportService.scala:92]:
```suggestion
      ts <-
        if (utcTs.truncatedTo(ChronoUnit.HOURS) == utcTs) Timestamp.fromInstant(utcTs.toInstant)
        else Left(s"Timestamp must be rounded to the hour: $utcTs")
```

**Review by cocreature — COMMENTED** [review]
haskell change loooks good, thanks!

**Review by mziolekda — APPROVED** [review]
Looks good in general. Left a couple of minor iems.
Please create a ticket to deal with the empty offsets in a follow-up PR

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala [hunk 2]
```diff
@@ -45,74 +73,111 @@ private[backend] object MeteringStorageBackendTemplate {
...
-        action_count,
-        metering_timestamp,
-        ledger_offset
+        sum(action_count)
       from transaction_metering
-      where ledger_offset <= ${ledgerEndCache()._1.toHexString.toString}
-      and   metering_timestamp >= ${from.micros}
-      and   (${isSet(to)} = 0 or metering_timestamp < ${to.map(_.micros)})
-      and   (${isSet(applicationId)} = 0 or application_id = ${applicationId.map(_.toString)})
+      where (${hasBegun(from)} = 0 or ledger_offset > $from)
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala:132]: `${hasBegun(from)} = 0 ` can be evaluated while constructing the sql statement.
Consider:
```
import com.daml.platform.store.backend.common.ComposableQuery.SqlStringInterpolation
val part: ComposableQuery.CompositeSql = if (!hasBegun(from)) cSQL"ledger_offset > $from" else cSQL""
...
SQL"... .$part ..."
```

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala [hunk 3]
```diff
@@ -45,74 +73,111 @@ private[backend] object MeteringStorageBackendTemplate {
...
+      }.toMap
+    }
+
+    ReportData(data, isFinal)
+
+  }
+
+  private def transactionMetering(
+      from: Offset,
+      to: Option[Time.Timestamp],
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala:123]: Maybe convey how the interval ends are treated in the parameter names:  `fromIncl/Excl` and `toIncl/Excl`

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala [hunk 4]
```diff
@@ -45,74 +73,111 @@ private[backend] object MeteringStorageBackendTemplate {
...
+      .asVectorOf(applicationCountParser)(connection)
+      .toMap
+
   }
 
-  /** Oracle does not understand true/false so compare against number
-    * @return 0 if the option is unset and non-zero otherwise
-    */
-  private def isSet(o: Option[_]): Int = o.fold(0)(_ => 1)
+  private def participantMetering(
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala:142]: Use a method names that start with a verb?

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/UpdateToMeteringDbDto.scala [hunk 5]
```diff
@@ -18,7 +18,7 @@ object UpdateToMeteringDbDto {
 
     if (input.nonEmpty) {
 
-      val ledgerOffset = input.head._1.toHexString
+      val ledgerOffset = input.last._1.toHexString
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/UpdateToMeteringDbDto.scala:21]: This was not a mistake originally, either the head or the last could be used (as they will both be within the same batch).  This makes more sense living alongside `participant_metering` which stores the last offset.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiMeteringReportService.scala [hunk 6]
```diff
@@ -85,25 +86,30 @@ private[apiserver] object ApiMeteringReportService {
...
+        if (utcTs.truncatedTo(ChronoUnit.HOURS) == utcTs) Right(())
+        else Left(s"Timestamp must be rounded to the hour: $utcTs")
+      ts <- Timestamp.fromInstant(utcTs.toInstant)
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiMeteringReportService.scala:95]: I've added this as a TODO.

**Comment by simonmaxen-da** [comment]
All review comments to be addressed as part of DPP-820.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala [hunk 7]
```diff
@@ -156,23 +149,27 @@ private[backend] trait StorageBackendTestsInitializeIngestion
...
 
     parties1 should have length 1
     packages1 should have size 1
     config1 shouldBe Some(offset(1) -> someConfiguration)
     contract41 should not be empty
     contract91 shouldBe None
     filterIds1 shouldBe List(1L, 4L) // since ledger-end does not limit the range query
-    metering1 should have length 2
+
+    // Metering can report include partially ingested data in non-final reports
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala:161]: done.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiMeteringReportService.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiMeteringReportService.scala:95]: done.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala (hunk 3 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala:123]: Added some scaladoc to explain the full purpose.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala (hunk 2 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala:132]: Done.

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala:161]: Done.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala (hunk 2 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala:132]: Done.

**Pull request merged**